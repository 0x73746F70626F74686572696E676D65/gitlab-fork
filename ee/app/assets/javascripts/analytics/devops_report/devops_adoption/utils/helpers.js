/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, and the check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, openModal }) => {
  if (openModal) {
    return false;
  } else if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );

  return anyPendingEnabledNamespaces;
};
