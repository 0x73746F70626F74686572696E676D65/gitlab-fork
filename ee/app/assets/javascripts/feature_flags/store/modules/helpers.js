// eslint-disable-next-line import/prefer-default-export
export const parseFeatureFlagsParams = params => ({
  operations_feature_flags: {
    name: params.name,
    description: params.description,
    active: true,
    scopes_attributes: params.scopes.map(scope => ({
      environment_scope: scope.name,
      active: scope.active,
    })),
  },
});
