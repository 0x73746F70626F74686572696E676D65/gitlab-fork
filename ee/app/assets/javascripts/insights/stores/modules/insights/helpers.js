import { CHART_TYPES } from 'ee/insights/constants';
import { __ } from '~/locale';

const getGroupByValue = (queryParams) => {
  return queryParams.group_by || queryParams.params.group_by;
};

const getTypeValue = (queryParams) => {
  return queryParams.issuable_type || queryParams.params.issuable_type;
};

const getAxisTitle = (label) => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: queryParams },
  { labels, datasets },
) => {
  const formattedData = {
    xAxisTitle: getAxisTitle(getGroupByValue(queryParams)),
    yAxisTitle: getAxisTitle(getTypeValue(queryParams)),
    labels,
    datasets: [],
    seriesNames: [],
  };

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: labels.map((label, i) => [label, dataset.data[i]]),
        })),
      );

      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
