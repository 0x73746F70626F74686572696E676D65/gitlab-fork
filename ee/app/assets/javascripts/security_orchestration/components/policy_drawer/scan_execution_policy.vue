<script>
import { GlSprintf } from '@gitlab/ui';
import { s__ } from '~/locale';
import {
  fromYaml,
  humanizeActions,
  humanizeRules,
} from '../policy_editor/scan_execution_policy/lib';
import { ACTIONS } from '../policy_editor/constants';
import { SUMMARY_TITLE } from './constants';
import PolicyDrawerLayout from './policy_drawer_layout.vue';
import PolicyInfoRow from './policy_info_row.vue';
import PolicyDrawerTags from './humanized_actions/tags.vue';
import PolicyDrawerVariables from './humanized_actions/variables.vue';

export default {
  i18n: {
    noActionMessage: s__('SecurityOrchestration|No actions defined - policy will not run.'),
    scanExecution: s__('SecurityOrchestration|Scan execution'),
    summary: SUMMARY_TITLE,
    ruleMessage: s__('SecurityOrchestration|And scans to be performed:'),
  },
  HUMANIZED_ACTION_COMPONENTS: {
    [ACTIONS.tags]: PolicyDrawerTags,
    [ACTIONS.variables]: PolicyDrawerVariables,
  },
  components: {
    PolicyDrawerTags,
    PolicyDrawerVariables,
    GlSprintf,
    PolicyDrawerLayout,
    PolicyInfoRow,
  },
  props: {
    policy: {
      type: Object,
      required: true,
    },
  },
  computed: {
    humanizedActions() {
      return humanizeActions(this.parsedYaml.actions);
    },
    humanizedRules() {
      return humanizeRules(this.parsedYaml.rules);
    },
    parsedYaml() {
      try {
        return fromYaml({ manifest: this.policy.yaml });
      } catch (e) {
        return null;
      }
    },
  },
  methods: {
    humanizedActionComponent({ action }) {
      return this.$options.HUMANIZED_ACTION_COMPONENTS[action];
    },
  },
};
</script>

<template>
  <policy-drawer-layout
    key="scan_execution_policy"
    :description="parsedYaml.description"
    :policy="policy"
    :type="$options.i18n.scanExecution"
  >
    <template v-if="parsedYaml" #summary>
      <policy-info-row data-testid="policy-summary" :label="$options.i18n.summary">
        <template v-if="!humanizedActions.length">{{ $options.i18n.noActionMessage }}</template>
        <div v-for="{ message, criteriaList } in humanizedActions" :key="message" class="gl-mb-3">
          <gl-sprintf :message="message">
            <template #scanner="{ content }">
              <strong>{{ content }}</strong>
            </template>
          </gl-sprintf>
          <ul class="gl-pl-6">
            <li v-for="criteria in criteriaList" :key="criteria.message" class="gl-mt-3">
              {{ criteria.message }}
              <component :is="humanizedActionComponent(criteria)" :criteria="criteria" />
            </li>
          </ul>
        </div>
        <div class="gl-mb-3">{{ $options.i18n.ruleMessage }}</div>
        <ul>
          <li v-for="(rule, idx) in humanizedRules" :key="idx">
            {{ rule }}
          </li>
        </ul>
      </policy-info-row>
    </template>
  </policy-drawer-layout>
</template>
