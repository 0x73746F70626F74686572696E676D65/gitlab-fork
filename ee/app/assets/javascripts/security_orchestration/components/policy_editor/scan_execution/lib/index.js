export { createPolicyObject, fromYaml } from './from_yaml';
export { toYaml } from './to_yaml';
export * from './rules';
export * from './cron';
export * from './actions';

export const DEFAULT_PROJECT_SCAN_EXECUTION_POLICY = `type: scan_execution_policy
name: ''
description: ''
enabled: true
rules:
  - type: pipeline
    branches:
      - '*'
actions:
  - scan: secret_detection
`;

export const DEFAULT_GROUP_SCAN_EXECUTION_POLICY = `type: scan_execution_policy
name: ''
description: ''
enabled: true
policy_scope:
  compliance_frameworks: []
rules:
  - type: pipeline
    branches:
      - '*'
actions:
  - scan: secret_detection
`;
