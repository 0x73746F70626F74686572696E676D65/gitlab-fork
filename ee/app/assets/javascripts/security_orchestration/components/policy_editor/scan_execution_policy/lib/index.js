export { fromYaml } from './from_yaml';
export { toYaml } from './to_yaml';
export * from './humanize';

export const DEFAULT_SCAN_EXECUTION_POLICY = `type: scan_execution_policy
name: # name is mandatory
description: ''
enabled: true
rules:
  - type: pipeline
    branches:
      - # branch is mandatory
actions:
  - scan: dast
    site_profile: ''
    scanner_profile: ''
`;
