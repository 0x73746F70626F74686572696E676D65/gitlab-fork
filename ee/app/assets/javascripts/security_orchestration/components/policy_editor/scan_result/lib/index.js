export { createPolicyObject, fromYaml } from './from_yaml';
export { toYaml } from './to_yaml';
export * from './rules';
export { approversOutOfSync, APPROVER_TYPE_DICT, APPROVER_TYPE_LIST_ITEMS } from './actions';
export * from './settings';
export * from './vulnerability_states';

export const DEFAULT_SCAN_RESULT_POLICY = `type: scan_result_policy
name: ''
description: ''
enabled: true
rules:
  - type: ''
actions:
  - type: require_approval
    approvals_required: 1
`;

export const SCAN_RESULT_POLICY_SETTINGS_POLICY = `type: scan_result_policy
name: ''
description: ''
enabled: true
rules:
  - type: ''
actions:
  - type: require_approval
    approvals_required: 1
approval_settings:
  block_protected_branch_modification:
    enabled: true
`;
