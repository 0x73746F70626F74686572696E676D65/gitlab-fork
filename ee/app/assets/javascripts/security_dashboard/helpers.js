import isPlainObject from 'lodash/isPlainObject';
import { s__ } from '~/locale';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { ALL, BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';

const parseOptions = obj =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = projects =>
  projects.map(p => ({ id: p.id.split('/').pop(), name: p.name }));

export const initFirstClassVulnerabilityFilters = projects => {
  const filters = [
    {
      name: s__('SecurityReports|Status'),
      id: 'state',
      options: [
        { id: ALL, name: s__('VulnerabilityStatusTypes|All') },
        ...parseOptions(VULNERABILITY_STATES),
      ],
      selection: new Set([ALL]),
    },
    {
      name: s__('SecurityReports|Severity'),
      id: 'severity',
      options: [BASE_FILTERS.severity, ...parseOptions(SEVERITY_LEVELS)],
      selection: new Set([ALL]),
    },
    {
      name: s__('SecurityReports|Report type'),
      id: 'reportType',
      options: [BASE_FILTERS.report_type, ...parseOptions(REPORT_TYPES)],
      selection: new Set([ALL]),
    },
  ];

  if (Array.isArray(projects)) {
    filters.push({
      name: s__('SecurityReports|Project'),
      id: 'projectId',
      options: [BASE_FILTERS.project_id, ...mapProjects(projects)],
      selection: new Set([ALL]),
    });
  }

  return filters;
};

export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter(entry => entry !== null);
};

export default () => ({});
