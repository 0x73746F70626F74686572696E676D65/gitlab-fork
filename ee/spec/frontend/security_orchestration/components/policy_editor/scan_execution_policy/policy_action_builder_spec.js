import { mountExtended } from 'helpers/vue_test_utils_helper';
import PolicyActionBuilder from 'ee/security_orchestration/components/policy_editor/scan_execution_policy/policy_action_builder.vue';
import { buildDefaultAction } from 'ee/security_orchestration/components/policy_editor/scan_execution_policy/lib';

describe('PolicyActionBuilder', () => {
  let wrapper;

  const factory = (options = {}) => {
    wrapper = mountExtended(PolicyActionBuilder, {
      propsData: {
        initAction: buildDefaultAction(),
        ...options,
      },
    });
  };

  beforeEach(() => {
    factory();
  });

  it('renders correctly with DAST as the default scanner', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
