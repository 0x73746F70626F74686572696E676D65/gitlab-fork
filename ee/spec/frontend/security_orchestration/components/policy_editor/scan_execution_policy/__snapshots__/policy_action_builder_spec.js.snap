// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyActionBuilder renders correctly with DAST as the default scanner 1`] = `
<div
  class="gl-bg-gray-10 gl-rounded-base gl-px-5! gl-pt-5! gl-relative gl-pb-4"
>
  <form
    class="form-inline"
  >
    <label
      class="text-uppercase gl-font-lg gl-mr-3"
      data-testid="action-component-label"
    >
      Then
    </label>
     Require a 
    <div
      class="dropdown b-dropdown gl-new-dropdown gl-mx-3 btn-group"
      data-testid="action-scanner-text"
      id="__BVID__6"
    >
      <!---->
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="btn dropdown-toggle btn-default btn-md gl-button gl-dropdown-toggle"
        id="__BVID__6__BV_toggle_"
        type="button"
      >
        <!---->
         
        <!---->
         
        <span
          class="gl-new-dropdown-button-text"
        >
          DAST
        </span>
         
        <svg
          aria-hidden="true"
          class="gl-button-icon dropdown-chevron gl-icon s16"
          data-testid="chevron-down-icon"
          role="img"
        >
          <use
            href="#chevron-down"
          />
        </svg>
      </button>
      <ul
        aria-labelledby="__BVID__6__BV_toggle_"
        class="dropdown-menu"
        role="menu"
        tabindex="-1"
      >
        <div
          class="gl-new-dropdown-inner"
        >
          <!---->
           
          <!---->
           
          <div
            class="gl-new-dropdown-contents"
          >
            <!---->
             
            <li
              class="gl-new-dropdown-item"
              role="presentation"
            >
              <button
                class="dropdown-item"
                role="menuitem"
                type="button"
              >
                <!---->
                 
                <!---->
                 
                <!---->
                 
                <div
                  class="gl-new-dropdown-item-text-wrapper"
                >
                  <p
                    class="gl-new-dropdown-item-text-primary"
                  >
                    
            SAST
          
                  </p>
                   
                  <!---->
                </div>
                 
                <!---->
              </button>
            </li>
            <li
              class="gl-new-dropdown-item"
              role="presentation"
            >
              <button
                class="dropdown-item"
                role="menuitem"
                type="button"
              >
                <!---->
                 
                <!---->
                 
                <!---->
                 
                <div
                  class="gl-new-dropdown-item-text-wrapper"
                >
                  <p
                    class="gl-new-dropdown-item-text-primary"
                  >
                    
            DAST
          
                  </p>
                   
                  <!---->
                </div>
                 
                <!---->
              </button>
            </li>
            <li
              class="gl-new-dropdown-item"
              role="presentation"
            >
              <button
                class="dropdown-item"
                role="menuitem"
                type="button"
              >
                <!---->
                 
                <!---->
                 
                <!---->
                 
                <div
                  class="gl-new-dropdown-item-text-wrapper"
                >
                  <p
                    class="gl-new-dropdown-item-text-primary"
                  >
                    
            Secret Detection
          
                  </p>
                   
                  <!---->
                </div>
                 
                <!---->
              </button>
            </li>
            <li
              class="gl-new-dropdown-item"
              role="presentation"
            >
              <button
                class="dropdown-item"
                role="menuitem"
                type="button"
              >
                <!---->
                 
                <!---->
                 
                <!---->
                 
                <div
                  class="gl-new-dropdown-item-text-wrapper"
                >
                  <p
                    class="gl-new-dropdown-item-text-primary"
                  >
                    
            Container Scanning
          
                  </p>
                   
                  <!---->
                </div>
                 
                <!---->
              </button>
            </li>
            <li
              class="gl-new-dropdown-item"
              role="presentation"
            >
              <button
                class="dropdown-item"
                role="menuitem"
                type="button"
              >
                <!---->
                 
                <!---->
                 
                <!---->
                 
                <div
                  class="gl-new-dropdown-item-text-wrapper"
                >
                  <p
                    class="gl-new-dropdown-item-text-primary"
                  >
                    
            Cluster Image Scanning
          
                  </p>
                   
                  <!---->
                </div>
                 
                <!---->
              </button>
            </li>
          </div>
           
          <!---->
        </div>
      </ul>
    </div>
     scan to run
  </form>
</div>
`;
